File: 98.validate-binary-search-tree.java
/*
 * @lc app=leetcode id=98 lang=java
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    // we need to consider edge cases of 2^31 and -2^31
    // and we can use long type to solve these cases
    public boolean isValidBST(TreeNode root, long lastMin, long lastMax) {
        if (root == null) {
            return true;
        }

        if (lastMin >= root.val || root.val >= lastMax) {
            return false;
        }
        
        return isValidBST(root.left, lastMin, root.val) && isValidBST(root.right, root.val, lastMax);
    }
}
// @lc code=end

File: 98.validate-binary-search-tree.cpp
/*
 * @lc app=leetcode id=98 lang=cpp
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Reference: https://leetcode.com/problems/validate-binary-search-tree/discuss/2066650/3-Approaches-Iterative-and-Recursive-C%2B%2B
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        // return recursive(root, LONG_MIN, LONG_MAX);
        return iterative(root);
    }

    bool iterative(TreeNode* root) {
        if (!root) {
            return true;
        }

        stack<TreeNode*> st;
        TreeNode* current = root;

        // previous node we pop out of the stack
        // for the BST, we only make sure if the previous node
        // is great than or equel to current, 
        // the BST is not valid
        TreeNode* previous = NULL;

        while (!st.empty() || current) {
            if (current) {
                st.push(current);
                current = current->left;
            } else {
                TreeNode* node = st.top(); st.pop();
                if (previous && previous->val >= node->val) {
                    return false;
                }

                previous = node;
                current = node->right;
            }
        }

        return true;
    }

    bool recursive(TreeNode* root, long min, long max) {
        if (!root) {
            return true;
        }

        // if we use int max or min, we can not handle edge cases 
        // of max and min here
        if (min >= root->val || root->val >= max) {
            return false;
        }

        return recursive(root->left, min, root->val) && recursive(root->right, root->val, max);
    }
};
// @lc code=end

